src/merkle/smt/mod.rs

```diff
@@ -140,6 +140,63 @@ pub(crate) trait SparseMerkleTree<const DEPTH: u8> {
         self.set_root(node_hash);
     }
 
+    /// Start a prospective mutation transaction, which can be queried, discarded, or applied.
+    ///
+    /// This method returns a [`MutationSet`], on which you can call [`MutationSet::insert()`] to
+    /// perform prospective mutations to this Merkle tree, and check the computed root hash given
+    /// those mutations with [`MutationSet::root()`]. When you are done, call
+    /// [`MutationSet::done()`], and pass that to [`Smt::commit()`] to apply the prospective
+    /// mutations to this tree. Or, to discard the chanes, simply [`drop`] the [`MutationSet`].
+    ///
+    /// ## Example
+    /// ```
+    /// # use miden_crypto::{hash::rpo::RpoDigest, Felt};
+    /// # use miden_crypto::merkle::{Smt, EmptySubtreeRoots, SMT_DEPTH};
+    /// # let mut smt = Smt::default();
+    /// let mut mutations = smt.mutate();
+    /// mutations.insert(Default::default(), Default::default());
+    /// assert_eq!(mutations.root(), *EmptySubtreeRoots::entry(SMT_DEPTH, 0));
+    /// smt.commit(mutations.done());
+    /// assert_eq!(smt.root(), *EmptySubtreeRoots::entry(SMT_DEPTH, 0));
+    /// ```
+    fn mutate(&self) -> MutationSet<DEPTH, Self>
+    where
+        Self: Sized,
+    {
+        MutationSet {
+            tree: self,
+            node_mutations: Default::default(),
+            new_pairs: Default::default(),
+            new_root: self.root(),
+        }
+    }
+
+    /// Apply the prospective mutations began with [`SparseMerkleTree::mutate()`] to this tree.
+    ///
+    /// This method takes the return value of [`MutationSet::done()`], and applies the changes
+    /// represented in that [`MutationSet`] to this Merkle tree. See [`SparseMerkleTree::mutate()`]
+    /// for more details.
+    fn commit(&mut self, mutations: CompletedMutationSet<DEPTH, Self::Key, Self::Value>)
+    where
+        Self: Sized,
+    {
```

Comment:
I'm wondering if we should return an error from here. Basically, a user could generate two different mutation sets starting from the same tree set. If then they apply the first set, applying the second set should fail.

Reply:
Yes I think we should. The borrow checker prevents the 'idiomatic' case:
```rust
let first = smt.mutate();
let second = smt.mutate();
smt.commit(first.done());
smt.commit(second.done()); // ERROR: cannot borrow `smt` as mutable because it is also borrowed as immutable
```

But if `second.done()` is pulled out into a variable beforehand first the borrow checker has no complaints, and if we simplify the interface as you suggested then `MutationSet` will no longer need to borrow `smt`.

That being said, even with the simplified interface we could instead have `MutationSet` *mutably* borrow `smt`, and have the borrow checker prevent generating two different mutation sets at the same time in the first place. Would you prefer that, or a fallible API?

Reply:
Let's go with the fallible API.